pip/pip-323.md

```diff
@@ -0,0 +1,167 @@
+get the usage of the quota (both are in seconds units).
+
+We will add the subscription name containing the oldest unacknowledged message to the Admin API
+topic stats endpoints (`{tenant}/{namespace}/{topic}/stats` and `{tenant}/{namespace}/{topic}/partitioned-stats`),
+allowing the user a complete workflow: alert using metrics when topic backlog is about to be exceeded, then
+query topic stats for that topic to retrieve the subscription name which contains the oldest message.
+For completeness, we will also add the backlog quota limits, both age and size, and the age of oldest 
+unacknowledged message.
+
+We will add a metric allowing the user to know how many times the usage exceeded the quota, both for time or size -
+`pulsar_storage_backlog_quota_exceeded_evictions_total`, where the `quota_type` label will be either `time` or 
+`size`. Monitoring that counter over time will allow the user to know when a topic backlog exceeded its quota,
+and if backlog eviction was chosen as action, then it happened, and how many times. 
+
+Some users may want the backlog quota check to happen more frequently, and as a consequence, the backlog age 
+metric more frequently updated. They can modify `backlogQuotaCheckIntervalInSeconds` configuration key, but without
+knowing how long this check takes, it will be hard for them. Hence, we will add the metric
+`pulsar_storage_backlog_quota_check_duration_seconds` which will be of histogram type.
+
+# Detailed Design
+
+## Public-facing Changes
+
+### Public API
+Adding the following to the response of topic stats, of both `{tenant}/{namespace}/{topic}/stats` 
+and `{tenant}/{namespace}/{topic}/partitioned-stats`:
+
+* `backlogQuotaLimitSize` - the size in bytes of the topic backlog quota 
+* `backlogQuotaLimitTime` - the topic backlog age quota, in seconds. 
+* `oldestBacklogMessageAgeSeconds` - the age of the oldest unacknowledged (i.e. backlog) message, measured by 
+   the time elapsed from its published time, in seconds. This value is recorded every backlog quota check 
+   interval, hence it represents the value seen in the last check.
+* `oldestBacklogMessageSubscriptionName` - the name of the subscription containing the oldest unacknowledged message.
+  This value is recorded every backlog quota check interval, hence it represents the value seen in the last check.
+
+
+### Metrics
+
+| Name                                                    | Description                                                                        | Attributes                        | Units   |
+|---------------------------------------------------------|------------------------------------------------------------------------------------|-----------------------------------|---------|
+| `pulsar_storage_backlog_age_seconds`                    | Gauge. The age of the oldest unacknowledged message (backlog)                      | cluster, topic                    | seconds |
```

Comment:
Do we need to mention that this one is only available if you enables the topic level metrics? For namespaces and broker level, we will not support this metric.

Reply:
Do we also need to add the namespace label even if we have the topic label? Usually, if you check the dashboard from Grafana, you will select the namespace first and then select the topic.